core/pkg/artifacts/multipart.go

```diff
@@ -0,0 +1,207 @@
+	partsInfo := make([]gql.UploadPartsInput, numParts)
+
+	startTime := time.Now()
+	ctx := context.Background() // TODO: Refactor ArtifactSaver to pass in a context for all the long running operations
+	g, ctx := errgroup.WithContext(ctx)
+	for i, task := range workerTasks {
+		g.Go(func() error {
+			worker := multipartHashWorker{
+				id:        i,
+				path:      path,
+				chunkSize: chunkSize,
+				fileSize:  fileSize,
+			}
+			return worker.hashFileParts(ctx, task, partsInfo)
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+
+	hashTime := time.Since(startTime)
+	hashSpeedMBps := float64(fileSize) / (1024 * 1024) / hashTime.Seconds()
+	logger.Debug("Computed multipart hashes",
+		"hashTimeMs", hashTime.Milliseconds(),
+		"hashSpeedMBps", hashSpeedMBps,
+		"numWorkers", numWorkers,
+		"numParts", len(partsInfo),
+		"fileSize", fileSize,
+		"chunkSize", chunkSize,
+	)
+	return partsInfo, nil
+}
+
+type hashWorkerTask struct {
+	startPart int // inclusive
+	endPart   int // exclusive
+}
+
+// splitHashTasks distributes parts evenly among workers for parallel processing.
+// If parts cannot be divided evenly, extra parts are distributed to earlier workers.
+func splitHashTasks(numParts int, numWorkers int) []hashWorkerTask {
+	partsPerWorker := numParts / numWorkers
+	workersWithOneMorePart := numParts % numWorkers
+
+	var workerTasks []hashWorkerTask
+	// NOTE: We start from 0, and only switch to 1-indexed when assiging the partNumber for server request
+	for i := range numWorkers {
+		var start, end int
+		if i < workersWithOneMorePart {
+			start = (partsPerWorker + 1) * i // All the previous workers get one more part
+			end = start + partsPerWorker + 1
+		} else {
+			start = workersWithOneMorePart + partsPerWorker*i
+			end = start + partsPerWorker
+		}
+		workerTasks = append(workerTasks, hashWorkerTask{
+			startPart: start,
+			endPart:   end,
+		})
+	}
+	return workerTasks
+}
+
+type multipartHashWorker struct {
+	id        int
+	path      string
+	chunkSize int64
+	fileSize  int64
+}
+
+// hashFileParts hash the md5 of consecutive parts of the file in serial.
+func (worker *multipartHashWorker) hashFileParts(
+	ctx context.Context,
+	task hashWorkerTask,
+	partsInfo []gql.UploadPartsInput,
+) error {
+	// One open file per worker so they seek in sequential order within its own file handler
+	// instead of jumping around by different workers. Might make file system's life easier (not benchmarked).
+	file, err := os.Open(worker.path)
+	if err != nil {
+		return fmt.Errorf("worker %d failed to open file: %w", worker.id, err)
+	}
+	//nolint:errcheck
+	defer file.Close()
+
+	for part := task.startPart; part < task.endPart; part++ {
+		// Return early if other worker had error
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		offset := int64(part) * worker.chunkSize
+
+		partSize := worker.chunkSize
+		if offset+partSize > worker.fileSize {
+			partSize = worker.fileSize - offset
+		}
```

Comment:
```suggestion
		partSize := min(worker.chunkSize, worker.fileSize - offset)
```

is equivalent right?

Reply:
Yep, changed to main.